k9mail/src/main/java/com/fsck/k9/controller/MessagingController.java

```diff
@@ -726,899 +714,65 @@ public void run() {
-                    account.getDescription(),
-                    folder,
-                    System.currentTimeMillis(),
-                    newMessages);
-
-            for (MessagingListener l : getListeners(listener)) {
-                l.synchronizeMailboxFinished(account, folder, remoteMessageCount, newMessages);
-            }
-
-
-            if (commandException != null) {
-                String rootMessage = getRootCauseMessage(commandException);
-                Timber.e("Root cause failure in %s:%s was '%s'",
-                        account.getDescription(), tLocalFolder.getName(), rootMessage);
-                localFolder.setStatus(rootMessage);
-                for (MessagingListener l : getListeners(listener)) {
-                    l.synchronizeMailboxFailed(account, folder, rootMessage);
-                }
-            }
-
-            Timber.i("Done synchronizing folder %s:%s", account.getDescription(), folder);
-
-        } catch (AuthenticationFailedException e) {
-            handleAuthenticationFailure(account, true);
-
-            for (MessagingListener l : getListeners(listener)) {
-                l.synchronizeMailboxFailed(account, folder, "Authentication failure");
-            }
-        } catch (Exception e) {
-            Timber.e(e, "synchronizeMailbox");
-            // If we don't set the last checked, it can try too often during
-            // failure conditions
-            String rootMessage = getRootCauseMessage(e);
-            if (tLocalFolder != null) {
-                try {
-                    tLocalFolder.setStatus(rootMessage);
-                    tLocalFolder.setLastChecked(System.currentTimeMillis());
-                } catch (MessagingException me) {
-                    Timber.e(e, "Could not set last checked on folder %s:%s",
-                            account.getDescription(), tLocalFolder.getName());
-                }
-            }
-
-            for (MessagingListener l : getListeners(listener)) {
-                l.synchronizeMailboxFailed(account, folder, rootMessage);
-            }
-            notifyUserIfCertificateProblem(account, e, true);
-            addErrorMessage(account, null, e);
-            Timber.e("Failed synchronizing folder %s:%s @ %tc", account.getDescription(), folder,
-                    System.currentTimeMillis());
-
-        } finally {
-            if (providedRemoteFolder == null) {
-                closeFolder(remoteFolder);
-            }
-
-            closeFolder(tLocalFolder);
+        String storeUri = account.getStoreUri();
+        if (ImapStore.isStoreUriImap(storeUri)) {
+            ImapSyncInteractor syncInteractor = new ImapSyncInteractor(account, folderName, listener, this);
+            syncInteractor.performSync(flagSyncHelper, messageDownloader, syncHelper);
+        } else {
+            LegacySyncInteractor.performSync(account, folderName, listener, this, messageDownloader, syncHelper);
```

Comment:
It's only use case is for IMAP IDLE. I think I removed that prematurely for when IDLE will get it's own class. I'll move it back.